presto-hive/src/main/java/com/facebook/presto/hive/metastore/ThriftHiveMetastore.java

```diff
@@ -564,9 +569,34 @@ public void alterPartition(String databaseName, String tableName, Partition part
     }
 
     @Override
-    public void grantTablePrivileges(String databaseName, String tableName, String grantee, Set<PrivilegeGrantInfo> privilegeGrantInfoSet)
+    public void grantTablePrivileges(String databaseName, String tableName, String grantee, Set<PrivilegeGrantInfo> requestedPrivileges)
     {
+        checkArgument(!containsAllPrivilege(requestedPrivileges), "\"ALL\" not supported in PrivilegeGrantInfo.privilege");
+
         try {
+            Set<HivePrivilegeInfo> existingPrivileges = getTablePrivileges(grantee, databaseName, tableName);
+
+            for (Iterator<PrivilegeGrantInfo> iterator = requestedPrivileges.iterator(); iterator.hasNext(); ) {
+                HivePrivilegeInfo requestedPrivilege = getOnlyElement(parsePrivilege(iterator.next()));
+
+                for (HivePrivilegeInfo existingPrivilege : existingPrivileges) {
+                    if ((requestedPrivilege.isContainedIn(existingPrivilege))) {
+                        iterator.remove();
```

Comment:
It's not a good idea to modify the set that's passed in:
1. This could fail if the set is immutable
2. If the caller continues to use the set after the call, it will cause subtle corruption bugs.